src/optimizer/rule/PushFilterDownAggregateRule.cpp

```diff
@@ -0,0 +1,125 @@
+    return pattern;
+}
+
+StatusOr<OptRule::TransformResult> PushFilterDownAggregateRule::transform(OptContext *octx,
+                                                       const MatchedResult &matched) const {
+    auto qctx = octx->qctx();
+    auto* filterGroupNode = matched.node;
+    auto* oldFilterNode = filterGroupNode->node();
+    auto deps = matched.dependencies;
+    DCHECK_EQ(deps.size(), 1);
+    auto aggGroupNode = deps.front().node;
+    auto* oldAggNode = aggGroupNode->node();
+    DCHECK(oldFilterNode->kind() == PlanNode::Kind::kFilter);
+    DCHECK(oldAggNode->kind() == PlanNode::Kind::kAggregate);
+    auto* newFilterNode =
+      const_cast<graph::Filter*>(static_cast<const graph::Filter*>(oldFilterNode))->clone(qctx);
+    auto* newAggNode =
+      const_cast<graph::Aggregate*>(static_cast<const graph::Aggregate*>(oldAggNode))->clone(qctx);
+    auto* condition = newFilterNode->condition();
+    auto& groupItems = newAggNode->groupItems();
+
+    // Check expression recursively to ensure no aggregate items in the filter
+    auto varProps = graph::ExpressionUtils::collectAll(condition,
+                                            {Expression::Kind::kVarProperty});
+    std::vector<std::string> propNames;
+    for (auto* expr : varProps) {
+        DCHECK(expr->kind() == Expression::Kind::kVarProperty);
+        propNames.emplace_back(*static_cast<const VariablePropertyExpression*>(expr)->prop());
+    }
+    std::unordered_map<std::string, Expression *> rewriteMap_;
+    bool noAggFilter = true;
+    auto colNames = newAggNode->colNames();
+    for (size_t i = 0; i < colNames.size(); ++i) {
+        auto& colName = colNames[i];
+        for (size_t j = 0; j < propNames.size(); ++j) {
+            if (!colName.compare(propNames[j])) {
+                if (graph::ExpressionUtils::findAny(groupItems[i],
+                                                   {Expression::Kind::kAggregate})) {
+                    noAggFilter = false;
+                    break;
+                } else {
+                    rewriteMap_[colName] = groupItems[i];
+                }
+            }
+        }
+        if (!noAggFilter) break;
+    }
+    if (!noAggFilter) {
+        // TODO: support partial push-down of filter
+        return TransformResult::noTransform();
+    }
+
+    // Rewrite VariablePropertyExpr in filter's condition
+    for (const auto& item : rewriteMap_) {
+        const VariablePropertyExpression rewrittenSubExpr(new std::string(),
+                                                          new std::string(item.first));
+        auto* rewriteSubExpr = item.second;
+        graph::RewriteAllSubExprVisitor
+               rewriteConditionVisitor(&rewrittenSubExpr, rewriteSubExpr, condition);
+        condition->accept(&rewriteConditionVisitor);
+        if (rewriteConditionVisitor.needRewriteRoot()) {
+            condition = qctx->objPool()->add(rewriteSubExpr->clone().release());
```

Comment:
I still don't find where the `condition` is used in the following parts of the `transform` function. So why to assign to it here ?

Reply:
`condition` is in newFilterNode.